### PR TITLE
Add supertype tests and add TSAbstractClassProperty

### PR DESCRIPTION
**Pull request by azz**
Thanks for the fix on the parser side, @soda0289 (eslint/typescript-eslint-parser#234)

* Super type fix: eslint/typescript-eslint-parser#242

#1306, #1422

### CODE REVIEW

**Comment by azz** [comment]
Build can be fixed when eslint/typescript-eslint-parser#244 is fixed.

File: src/printer.js [hunk 1]
```diff
@@ -1447,29 +1447,30 @@ function genericPrintNoParens(path, options, print, args) {
...
       if (n.static) parts.push("static ");
 
-      var key;
+      var key = [];
+
+      var variance = getFlowVariance(n, options);
+
+      if (variance) {
+        key.push(variance);
+      } else if (n.accessibility) {
```
Comment (vjeux) [inline src/printer.js]: what happens if you put `if` on its own and not behind `else`?

Comment (azz) [inline src/printer.js]: I don't think `n.accessibility` is ever set in flow, so it should be ok to break it out. Tests still pass.

**Comment by vjeux** [comment]
Yay, thanks!

Do you have a sense of how close we are to being complete enough so that most typescript programs are able to be printed?

**Comment by azz** [comment]
@vjeux Interfaces, modules, and enums are still missing I think. Once those are done we should be printing _most_ of the syntax. Then we can copy across the entire test suite from the TypeScript repo, see if any parse errors occur. 

There will likely be a number of bugs due to the (frankly _insane_) amount of syntax that TypeScript has, on top of an already syntax-heavy language.

**Comment by vjeux** [comment]
Thanks, sounds like a plan :)

And yeah, the amount of syntax of JS + ES6 + Flow + JSX + TypeScript is crazy :p

**Comment by despairblue** [comment]
Yeah, I keep stumbling into things as well `declare keyword`, `module block`, `enum member`...list goes on :smile:

**Comment by despairblue** [comment]
> @vjeux Interfaces, modules, and enums are still missing I think. 

All in #1459 now. There gonna be a lot of warts since I skipped over pretty much all `modifiers` etc. but it should print. I'll try rerunning this over the TypeScript test suite.

**Pull request merged**